File: code/cpp/src/luogu/done/p1776.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    int i, j, n, w, vv, ww, mm, count, vvv, www;
    cin >> n >> w;

    vector<int> dp(1 + w, 0);

    for (i = 0; i < n; ++i) {
        vv, ww, mm;
        cin >> vv >> ww >> mm;
        count = 1;
        while (mm > count) {
            mm -= count;
            vvv = vv * count;
            www = ww * count;
            for (j = w; j >= www; --j) {
                dp[j] = max(dp[j], dp[j - www] + vvv);
            }
            count <<= 1;
        }
        vvv = vv * mm;
        www = ww * mm;
        for (j = w; j >= www; --j) {
            dp[j] = max(dp[j], dp[j - www] + vvv);
        }
    }

    cout << dp[w] << endl;

    return 0;
}

File: code/py/src/onlinejudge/luogu/unsolved/p1776.py
def main():
    n, w = map(int, input().split())

    dp = [0 for _ in range(1 + w)]

    for _ in range(n):
        vv, ww, mm = map(int, input().split())
        count = 1
        while mm > count:
            mm -= count
            vvv = vv * count
            www = ww * count
            for j in range(w, www - 1, -1):
                dp[j] = max(dp[j], dp[j - www] + vvv)
            count <<= 1
        vvv = vv * mm
        www = ww * mm
        for j in range(w, www - 1, -1):
            dp[j] = max(dp[j], dp[j - www] + vvv)

    print(dp[w])


if __name__ == '__main__':
    main()
